sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Misc/MiscellaneousTests.cs

```diff
@@ -0,0 +1,245 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2019 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+extern alias csharp;
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using System.Threading;
+using csharp::SonarAnalyzer.Rules.CSharp;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using SonarAnalyzer.Helpers;
+using SonarAnalyzer.UnitTest.TestFramework;
+
+namespace SonarAnalyzer.UnitTest.Misc
+{
+    [TestClass]
+    public class MiscellaneousTests
+    {
+        public TestContext TestContext { get; set; }
+
+        [TestMethod]
+        // This test takes around 40s-60s to run, which is as long as all of the other tests.
+        // This category allows us to exclude it from the set of tests to run.
+        // In the IDE: set the test filter in the Test Explorer to the following
+        //      -Trait:"Slow"
+        // dotnet test or vstest.console.exe: pass the following command line argument
+        //       --TestCaseFilter:"TestCategory!=Slow"
+        [TestCategory("Slow")]
+        public void TestForSlowAnalyzers_EFMigrationSource_2474()
+        {
+            // The test file comes from a repro project for https://github.com/SonarSource/sonar-dotnet/issues/2474
+            // The single file is enough to cause the issue described in the bug, although there
+            // are other large files in the project that might also be problematic.
+
+            // Analyzers known to timeout against this test code
+            // (note: this test doesn't cover the utility analyzers):
+            var knownSlowAnalyzers = new[] { "SonarAnalyzer.Rules.CSharp.UnusedPrivateMember" };
+            
+            int analysisTimeoutInMs = 700;
```

Comment:
Any analyser that takes longer than 700ms will be aborted.
To stop the test from failing continually, there's a list of analyzers that are known to be slow.

The test will fail if either:
* any of the aborted analyzers are not in the "known slow analyzer" list, or
* a "known slow analyzer" is not in the list (i.e. that analyzer is now faster so the "slow list" needs to be updated)

This test is just a starting point, created to help diagnose #2474. It was enough to identify that only one of the "normal" (i.e. non-utility analysers) has a problem processing this file.